File: src/main/java/com/keecker/services/navigation/interfaces/DetectionResult.java
package com.keecker.services.navigation.interfaces;

import android.os.Parcel;
import android.os.Parcelable;

import com.keecker.services.utils.geometry.Quaternion;
import com.keecker.services.utils.geometry.Transform;
import com.keecker.services.utils.geometry.Vector3;

import java.util.Arrays;

/**
 * Created by xavier on 12/02/18.
 */
public class DetectionResult implements Parcelable{
    public final Transform transform;
    public final double confidence;
    public final double timestamp;

    protected DetectionResult(Parcel in) {
        transform = in.readParcelable(Transform.class.getClassLoader());
        confidence = in.readDouble();
        timestamp = in.readDouble();
    }

    public static final Creator<DetectionResult> CREATOR = new Creator<DetectionResult>() {
        @Override
        public DetectionResult createFromParcel(Parcel in) {
            return new DetectionResult(in);
        }

        @Override
        public DetectionResult[] newArray(int size) {
            return new DetectionResult[size];
        }
    };

    @Override
    public String toString() {
        return "DetectionResult{" +
                "transform=" + transform +
                ", confidence=" + confidence +
                ", timestamp=" + timestamp +
                '}';
    }

    public DetectionResult(float result[], double confidence, double timestamp) {
        this.transform = new Transform(
            new Vector3(result[0], result[1], result[2]),
            Quaternion.fromTheta(result[3])
        );
        this.confidence = confidence;
        this.timestamp = timestamp;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(transform, flags);
        dest.writeDouble(confidence);
        dest.writeDouble(timestamp);
    }
}

File: src/main/java/com/keecker/services/navigation/interfaces/DetectionResult.kt
/*
 *  Copyright (C) 2018 KEECKER SAS (www.keecker.com)
 *  Unauthorized copying of this file, via any medium is strictly prohibited
 *  Proprietary and confidential
 *
 *  Created by Emeric Colombe<emeric@keecker.com>
 *
 */
package com.keecker.services.navigation.interfaces

import android.graphics.RectF
import android.os.Parcel
import android.os.Parcelable
import com.keecker.embedded.camera.interfaces.Frame

data class DetectionResult (
        var id : String,
        var title : String,
        var confidence : Float,
        var location : RectF,
        var frame: Frame? = null) : Parcelable {

    constructor(parcel: Parcel)
            : this(
            parcel.readString(),
            parcel.readString(),
            parcel.readFloat(),
            parcel.readParcelable(RectF::class.java.classLoader),
            parcel.readParcelable(Frame::class.java.classLoader))

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(this.id)
        dest.writeString(this.title)
        dest.writeFloat(this.confidence)
        dest.writeParcelable(this.location, flags)
        dest.writeParcelable(this.frame, flags)
    }


    companion object CREATOR : Parcelable.Creator<DetectionResult> {
        override fun createFromParcel(parcel: Parcel): DetectionResult {
            return DetectionResult(parcel)
        }

        override fun newArray(size: Int): Array<DetectionResult?> {
            return arrayOfNulls(size)
        }
    }

}
